Mathlib/LinearAlgebra/RootSystem/Finite/CanonicalBilinear.lean

```diff
@@ -0,0 +1,139 @@
+positive-definite on the span of roots.
+From these facts, it is easy to show that Coxeter weights in a finite root pairing are bounded
+above by 4. Thus, the pairings of roots and coroots in a crystallographic root pairing are
+restricted to a small finite set of possibilities.
+Another application is to the faithfulness of the Weyl group action on roots, and finiteness of the
+Weyl group.
+
+## Main definitions:
+ * `Polarization`: A distinguished linear map from the weight space to the coweight space.
+ * `CanonicalBilinear` : The bilinear form on weight space corresponding to `Polarization`.
+
+## References:
+ * SGAIII Exp. XXI
+ * Bourbaki, Lie groups and Lie algebras
+
+## Main results:
+ * `polarization_self_sum_of_squares` : The inner product of any weight vector is a sum of squares.
+ * `canonicalBilinear_reflection_reflection_apply` : `CanonicalBilinear` is invariant with respect
+   to reflections.
+ * `canonicalBilinear_self_smul_coroot`: The inner product of a root with itself times the
+   corresponding coroot is equal to two times Polarization applied to the root.
+
+## TODO (possibly in other files)
+ * Positivity and nondegeneracy
+ * Weyl-invariance
+ * Faithfulness of Weyl group action, and finiteness of Weyl group, for finite root systems.
+ * Relation to Coxeter weight.  In particular, positivity constraints for finite root pairings mean
+  we restrict to weights between 0 and 4.
+-/
+
+open Set Function
+open Module hiding reflection
+
+noncomputable section
+
+variable {ι R M N : Type*}
+
+lemma isSumSq_of_sum_of_squares [Mul R] [AddCommMonoid R] (s : Finset ι) (f : ι → R) :
+    IsSumSq (∑ i ∈ s, f i * f i) := by
+  induction s using Finset.cons_induction with
+  | empty =>
+    simpa only [Finset.sum_empty] using IsSumSq.zero
+  | cons i s his h =>
+    simp only [Finset.sum_cons]
+    exact IsSumSq.sq_add (f i) (∑ i ∈ s, f i * f i) h
+--#find_home! isSumSq_of_sum_of_squares --returns this file
+
+namespace RootPairing
+
+section CommRing
+
+variable [Fintype ι] [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+(P : RootPairing ι R M N)
+
+/-- An invariant linear map from weight space to coweight space. -/
+@[simps]
+def Polarization : M →ₗ[R] N where
+  toFun m := ∑ (i : ι), P.toPerfectPairing m (P.coroot i) • (P.coroot i)
+  map_add' x y := by
+    simp only [← toLin_toPerfectPairing, map_add, PerfectPairing.toLin_apply, LinearMap.add_apply,
+      add_smul, Finset.sum_add_distrib]
+  map_smul' r x := by
+    simp only [← toLin_toPerfectPairing, map_smul, LinearMap.smul_apply, RingHom.id_apply,
+      Finset.smul_sum, smul_assoc]
+
+/-- An invariant inner product on the weight space. -/
+@[simps]
+def CanonicalBilinear : M →ₗ[R] M →ₗ[R] R where
+  toFun x := P.toLin x ∘ₗ P.Polarization
+  map_add' x y := by simp only [map_add, LinearMap.add_comp]
+  map_smul' r x := by simp only [LinearMapClass.map_smul, RingHom.id_apply, LinearMap.smul_comp]
```

Comment:
Again, we can avoid proving linearity by hand. What do you think about this naming?
```suggestion
def RootForm : LinearMap.BilinForm R M :=
  ∑ i, (LinearMap.lsmul R R).compl₁₂ (P.coroot' i) (P.coroot' i)

/-- An invariant inner product on the coweight space. -/
def CorootForm : LinearMap.BilinForm R N :=
  ∑ i, (LinearMap.lsmul R R).compl₁₂ (P.root' i) (P.root' i)
```